editor/src/clj/util/http_client.clj

```diff
@@ -13,103 +13,56 @@
-    (when (and content-type (not character-encoding))
-      (.setRequestProperty conn "Content-Type" content-type))
-    (doseq [[h v] headers]
-      (.setRequestProperty conn h v))
-    (when (false? follow-redirects)
-      (.setInstanceFollowRedirects ^HttpURLConnection conn false))
-    (.setRequestMethod ^HttpURLConnection conn (.toUpperCase (name request-method)))
-    (when body
-      (.setDoOutput conn true))
-    (when socket-timeout
-      (.setReadTimeout conn socket-timeout))
-    (when conn-timeout
-      (.setConnectTimeout conn conn-timeout))
-    (when chunk-size
-      (.setChunkedStreamingMode conn chunk-size))
-    (.connect conn)
-    (when body
-      (with-open [out (.getOutputStream conn)]
-        (io/copy body out)))
-    (merge {:headers (parse-headers conn)
-            :status (.getResponseCode ^HttpURLConnection conn)
-            :body (when-not (= request-method :head)
-                    (coerce-body-entity req conn))}
-           (when save-request?
-             {:request (assoc (dissoc req :save-request?)
-                              :http-url http-url)}))))
+  Returns a CompletableFuture that will receive a map with the following keys:
+    :status     HTTP status code, e.g. 200
+    :headers    response headers map, string->(string|string[])
+    :body       the key is present only if :as was provided; either string,
+                InputStream or byte array"
+  [url & {:keys [method body headers as]
+          :or {method "GET"}}]
+  {:pre [(string? url)]}
+  (-> client
+      (.sendAsync
+        (-> (HttpRequest/newBuilder (URI. url))
+            (as-> $ (reduce-kv HttpRequest$Builder/.header $ headers))
+            (.method method (cond
+                              (nil? body) (HttpRequest$BodyPublishers/noBody)
+                              (string? body) (HttpRequest$BodyPublishers/ofString body)
+                              (instance? InputStream body) (HttpRequest$BodyPublishers/ofInputStream (fn [] body))
+                              (instance? byte/1 body) (HttpRequest$BodyPublishers/ofByteArray body)
+                              :else (throw (IllegalArgumentException. "body should either be nil, string, InputStream or byte array"))))
+            (.build))
+        (case as
+          nil (HttpResponse$BodyHandlers/discarding)
+          :string (HttpResponse$BodyHandlers/ofString)
+          :input-stream (HttpResponse$BodyHandlers/ofInputStream)
+          :byte-array (HttpResponse$BodyHandlers/ofByteArray)))
+      (.thenApply
+        (fn [^HttpResponse response]
+          (cond-> {:status (.statusCode response)
+                   :headers (into {}
+                                  (map (fn [[k v]]
+                                         [k (if (< 1 (count v)) (vec v) (.get ^List v 0))]))
+                                  (.map (.headers response)))}
+                  (.body response)
+                  (assoc :body (.body response)))))))
```

Comment:
I rewrote the HTTP client to use a new `java.net.http` API (since JDK 11). The main reason was that I wanted to use async HTTP requests for better editor scripts performance.